lib/active_merchant/billing/gateways/moneris.rb

```diff
@@ -208,6 +208,52 @@ def add_cof(post, options)
         post[:payment_information] = options[:payment_information] if options[:payment_information]
       end
 
+      def add_stored_credential(post, options)
+        add_cof(post, options)
+        # if any of :issuer_id, :payment_information, or :payment_indicator is not passed,
+        # then check for :stored credential options
+        return unless (stored_credential = options[:stored_credential]) && cof_details_not_present?(options)
+        if stored_credential[:initial_transaction]
+          add_stored_credential_initial(post, options)
+        else
+          add_stored_credential_used(post, options)
+        end
+      end
+
+      def cof_details_not_present?(options)
+        options.dig(:issuer_id).nil? || options.dig(:payment_indicator).nil? || options.dig(:payment_information).nil?
+      end
+
+      def add_stored_credential_initial(post, options)
+        post[:payment_information] ||= '0'
+        post[:issuer_id] ||= ''
+        if options[:stored_credential][:initiator] == 'merchant'
+          case options[:stored_credential][:reason_type]
+          when 'recurring', 'installment'
+            post[:payment_indicator] ||= 'R'
+          when 'unscheduled'
+            post[:payment_indicator] ||= 'C'
+          end
+        else
+          post[:payment_indicator] ||= 'C'
+        end
+      end
+
+      def add_stored_credential_used(post, options)
+        post[:payment_information] ||= '2'
+        post[:issuer_id] = options[:stored_credential][:network_transaction_id] if post[:issuer_id] == ''
```

Comment:
Should this be `if options[:issuer_id] == ''`? If so, what about when `:issuer_id` is nil?

Reply:
Yes good catch, I should have used `options` instead of `post`. Thank you! `auth_purchase_params` in core will convert all nil `:issuer_id` to `''`. Hmm now I'm wondering if there's something else I'm failing to consider?

Reply:
Got it. Thanks for the tip. The changes have been pushed.